### PR TITLE
Better performance with visualization

### PR DESCRIPTION
**Pull request by DaddyWesker**
Working with [this PR](https://github.com/trueagi-io/minecraft-demo/pull/161) only.

### CODE REVIEW

File: src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java [hunk 1]
```diff
@@ -54,27 +54,54 @@ public int getWidth()
...
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        }
-        imageclass.close();
-        if (ni_bytes.length <= 0)
-            return null;
-        else {
-            return ByteBuffer.wrap(ni_bytes);
-        }
+        long time2 = System.nanoTime();
```
Comment (noskill) [inline src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: We can save performance stats using 
Profiler profiler = MinecraftClient.getInstance().world.getProfiler();

then you start profiling with:
profiler.push("getRGBFrame");

Comment (noskill) [reply src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: and stop with profiler.pop("getRGBFrame");

**Review by noskill — CHANGES_REQUESTED** [review]
I suggest to utilise minecraft profiler for profiling

**Comment by DaddyWesker** [comment]
@noskill current `long time2 = System.nanoTime();` is just a garbage i haven't deleted yet since this code in the testing stage. And i need you to launch a game and check if game on your side has statters issue on fullscreen or not.

**Comment by noskill** [comment]
there is no stuttering and the performance is much better

**Comment by DaddyWesker** [comment]
So it seems that i'm the only one experiencing stutters for some reason. Ok.

**Comment by noskill** [comment]
@DaddyWesker i used test_vereya.py script for testing. If you used different testing environment it might be the reason why test results are different.

**Comment by DaddyWesker** [comment]
@noskill well, i'm just overtook control and walked around myself. With fullscreen. No stuttering on lesser size of screen though.

**Comment by DaddyWesker** [comment]
Ok, this PR is now not Draft.

File: src/main/java/io/singularitynet/utils/ImageClass.java [hunk 2]
```diff
@@ -397,6 +402,13 @@ private void uploadInternal(int level, int offsetX, int offsetY, int unpackSkipP
         }
     }
 
+    public ByteBuffer getImageAsByteBuffer() {
+        ByteBuffer res = BufferUtils.createByteBuffer(this.getHeight() * this.getWidth() * 4);
+        res.flip();
+        GlStateManager._readPixels(0, 0, this.getWidth(), this.getHeight(), this.format.toGl(), GL11.GL_UNSIGNED_BYTE, res);
```
Comment (noskill) [inline src/main/java/io/singularitynet/utils/ImageClass.java]: It looks like ImageClass is not used anymore except Format class. I think we can safely remove it and use Format from NativeImage, or just keep the copy of Format, without ImageClass.

**Review by CICS-Oleg — APPROVED** [review]

File: src/main/java/io/singularitynet/utils/ImageClass.java (hunk 2 above)
Comment (DaddyWesker) [reply src/main/java/io/singularitynet/utils/ImageClass.java]: So, you're suggesting to remove ImageClass and extract getImageAsByteBuffer to  VideoProducerImplementation?

Comment (noskill) [reply src/main/java/io/singularitynet/utils/ImageClass.java]: Yes

Comment (DaddyWesker) [reply src/main/java/io/singularitynet/utils/ImageClass.java]: Done

File: src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java [hunk 3]
```diff
@@ -58,8 +60,10 @@ private AbstractMap.Entry<ByteBuffer, int[]> getRGBFrame()
         Framebuffer framebuffer = MinecraftClient.getInstance().getFramebuffer();
         int i = framebuffer.textureWidth;
         int j = framebuffer.textureHeight;
-        ImageClass imageclass = new ImageClass(i, j, false);
-        ByteBuffer resBuffer = imageclass.getImageAsByteBuffer();
+        NativeImage nativeimage = new NativeImage(i, j, false);
+        ByteBuffer resBuffer = BufferUtils.createByteBuffer(nativeimage.getHeight() * nativeimage.getWidth() * 4);
+        resBuffer.flip();
+        GlStateManager._readPixels(0, 0, nativeimage.getWidth(), nativeimage.getHeight(), nativeimage.getFormat().toGl(), GL11.GL_UNSIGNED_BYTE, resBuffer);
```
Comment (noskill) [inline src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]:
```java
GlStateManager._readPixels(0, 0, i, j, NativeImage.Format.RGBA.toGl(), GL11.GL_UNSIGNED_BYTE, resBuffer);
```

File: src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java [hunk 4]
```diff
@@ -58,8 +60,10 @@ private AbstractMap.Entry<ByteBuffer, int[]> getRGBFrame()
         Framebuffer framebuffer = MinecraftClient.getInstance().getFramebuffer();
         int i = framebuffer.textureWidth;
         int j = framebuffer.textureHeight;
-        ImageClass imageclass = new ImageClass(i, j, false);
-        ByteBuffer resBuffer = imageclass.getImageAsByteBuffer();
+        NativeImage nativeimage = new NativeImage(i, j, false);
```
Comment (noskill) [inline src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: nativeimage allocates some memory, which we don't use, i suggest removing it

**Review by noskill — CHANGES_REQUESTED** [review]

File: src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java (hunk 3 above)
Comment (noskill) [reply src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: Or RGB - maybe it will be faster

Comment (DaddyWesker) [reply src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: If I'm putting RGB there then game will crash each launch of an agent.

So, you want to just remove `NativeImage nativeimage = new NativeImage(i, j, false);` and its usage?

File: src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java (hunk 4 above)
Comment (DaddyWesker) [reply src/main/java/io/singularitynet/MissionHandlers/VideoProducerImplementation.java]: I'll see if that won't hurt image size on Mac.

**Comment by DaddyWesker** [comment]
@noskill Done

**Review by noskill — APPROVED** [review]

**Pull request merged**